flink-connectors/flink-connector-dynamodb/src/main/java/org/apache/flink/streaming/connectors/dynamodb/sink/DynamoDbSink.java

```diff
@@ -0,0 +1,140 @@
+ *       add elements will be blocked while the number of elements in the buffer is at the maximum
+ *   <li>{@code maxBatchSizeInBytes}: the maximum size of a batch of entries that may be written to
+ *       DynamoDb measured in bytes
+ *   <li>{@code maxTimeInBufferMS}: the maximum amount of time an entry is allowed to live in the
+ *       buffer, if any element reaches this age, the entire buffer will be flushed immediately
+ *   <li>{@code maxRecordSizeInBytes}: the maximum size of a record the sink will accept into the
+ *       buffer, a record of size larger than this will be rejected when passed to the sink
+ *   <li>{@code failOnError}: when an exception is encountered while persisting to DynamoDb, the job
+ *       will fail immediately if failOnError is set
+ *   <li>{@code dynamoDbTablesConfig}: if provided for the table, the DynamoDb sink will attempt to
+ *       deduplicate records with the same primary and/or secondary keys in the same batch request.
+ *       Only the latest record with the same combination of key attributes is preserved in the
+ *       request.
+ * </ul>
+ *
+ * <p>Please see the writer implementation in {@link DynamoDbSinkWriter}
+ *
+ * @param <InputT> Type of the elements handled by this sink
+ */
+@PublicEvolving
+public class DynamoDbSink<InputT> extends AsyncSinkBase<InputT, DynamoDbWriteRequest> {
+
+    private final Properties dynamoDbClientProperties;
+    private final DynamoDbTablesConfig dynamoDbTablesConfig;
+    private final boolean failOnError;
+
+    protected DynamoDbSink(
+            ElementConverter<InputT, DynamoDbWriteRequest> elementConverter,
+            int maxBatchSize,
+            int maxInFlightRequests,
+            int maxBufferedRequests,
+            long maxBatchSizeInBytes,
+            long maxTimeInBufferMS,
+            long maxRecordSizeInBytes,
+            boolean failOnError,
+            DynamoDbTablesConfig dynamoDbTablesConfig,
+            Properties dynamoDbClientProperties) {
+        super(
+                elementConverter,
+                maxBatchSize,
+                maxInFlightRequests,
+                maxBufferedRequests,
+                maxBatchSizeInBytes,
+                maxTimeInBufferMS,
+                maxRecordSizeInBytes);
+        checkNotNull(dynamoDbTablesConfig, "Missing dynamoDbTablesConfig");
+        checkNotNull(dynamoDbClientProperties, "Missing dynamoDbClientProperties");
+        this.failOnError = failOnError;
+        this.dynamoDbTablesConfig = dynamoDbTablesConfig;
+        this.dynamoDbClientProperties = dynamoDbClientProperties;
+    }
+
+    /**
+     * Create a {@link DynamoDbSinkBuilder} to construct a new {@link DynamoDbSink}.
+     *
+     * @param <InputT> type of incoming records
+     * @return {@link DynamoDbSinkBuilder}
+     */
+    public static <InputT> DynamoDbSinkBuilder<InputT> builder() {
+        return new DynamoDbSinkBuilder<>();
+    }
+
+    @Experimental
```

Comment:
I think `@Internal` might be more appropriate here, since we're not expecting to remove this method... `flink-architecture-tests` are cool with `@Internal` here too.

Reply:
Thank you @CrynetLogistics for the review. I wasn't sure about the documentation that is published on the website. I'll read up about the process and add it in this PR if possible

Reply:
Can you please give me a reference to the changes for the elements converter?

Reply:
> Can you please give me a reference to the changes for the elements converter?

Sure, here is the link to the [commit](https://github.com/apache/flink/commit/25ecc0b9202050b340fb4d6fc95b0585a8119937) on the [master branch](https://github.com/apache/flink/commits/master).

Reply:
I can see how it is useful for the KinesisSink, as we do not leak Kinesis Put request to the user, but I think it would be more complicated to implement for the DynamoDB Sink. 

We will have to implement something that would look exactly like DynamoDB WriteRequest with all its attributes, types, request types, etc. and translate it internally into the WriteRequest object (by the converter). I can not see any other solution for that.

It will be a lot of functionality we will have to "repeat" that dynamodb provides.

Do you think it worth doing it, instead of letting user to define the converter themselves? @CrynetLogistics

Reply:
FYI added https://issues.apache.org/jira/browse/FLINK-25859 for documentation

Reply:
@CrynetLogistics @dannycranmer apologize for ping. We have prepared a sample [PR](https://github.com/YuriGusev/flink/commit/de83346dc24367ff1101ae13934dfcf1fc649de2) in case hiding the ElementConverter is a requirement.